### ISSUE TITLE
HTTP_HOST not parsing port correctly at 2.X

### ISSUE THREAD

**Issue by caiofct** [rack/rack#2070]
Hello, I think we might have a regression from this PR https://github.com/rack/rack/pull/1606.

With Rack 2.1.4.3 (Rack 3.X also works as expected)

```ruby
irb(main):001:0> r = Rack::Request.new({"HTTP_HOST" => "some_service:3001"})
irb(main):002:0> r.host
=> "some_service"
irb(main):003:0> r.port
=> 3001
```

With Rack 2.2.6.4 (But any version at 2.2.X will behave the same way)
```ruby
irb(main):001:0> r = Rack::Request.new({ "HTTP_HOST" => "some_service:3001" })
irb(main):002:0> r.host
=> "http://some_service:3001"
irb(main):003:0> r.port
=> nil
```

Am I missing something here?

Ruby: 2.7.8
Rails 7.0.4

**Comment by jeremyevans** [comment]
Originally I thought this was a non issue, because your code for the failing example uses `"HTTP_HOST" => "http://some_service:3001"` (which is an invalid Host header as it includes the protocol).  However, when using `"HTTP_HOST" => "some_service:3001"`, the port is not split correctly from the host in Rack 2.2.

I checked some rack versions:

```
2.1.4.3, 3.0.3:  ["some_service", 3001]
2.2.0: ["some", nil]
2.2.2, 2.2.3, 2.2.3.1: ["some_service:3001", nil]
```

The issue is likely related to the use of `_` in the host names.  In DNS, `_` is not a valid hostname character.  However, for non-DNS host naming, `_` may be valid.

A potential fix would be adding `_` as a valid character here: https://github.com/rack/rack/blob/2-2-stable/lib/rack/request.rb#L611

**Comment by caiofct** [comment]
@jeremyevans Thanks for taking a look at it. The failing example with the protocol was my mistake. And as you tested it also fails without it. Also your point about valid dns hostnames make sense and I wasn't considering it. The some_service host comes from docker compose which is a service we have and using underscore there is common practice however we could very easily use a dash instead of underscore and that would solve the issue but I wonder if rack 2.2 behavior compared to 2.1 or even 3.0 is not a regression.

> Originally I thought this was a non issue, because your code for the failing example uses `"HTTP_HOST" => "http://some_service:3001"` (which is an invalid Host header as it includes the protocol). However, when using `"HTTP_HOST" => "some_service:3001"`, the port is not split correctly from the host in Rack 2.2.
> 
> I checked some rack versions:
> 
> ```
> 2.1.4.3, 3.0.3:  ["some_service", 3001]
> 2.2.0: ["some", nil]
> 2.2.2, 2.2.3, 2.2.3.1: ["some_service:3001", nil]
> ```
> 
> The issue is likely related to the use of `_` in the host names. In DNS, `_` is not a valid hostname character. However, for non-DNS host naming, `_` may be valid.
> 
> A potential fix would be adding `_` as a valid character here: https://github.com/rack/rack/blob/2-2-stable/lib/rack/request.rb#L611

**Comment by jeremyevans** [comment]
I submitted #2071 to change the Rack 2.2 behavior to align with Rack 2.1 and 3.0.  I'm not saying with the behavior is correct, but this at least makes it consistent.

**Comment by ioquatix** [comment]
The proposed fix was merged.

**Comment by ioquatix** [comment]
Released rack 2.2.7 with the fix.
